scipy/stats/morestats.py

```diff
@@ -1215,28 +1253,337 @@ def boxcox_normplot(x, la, lb, plot=None, N=80):
     >>> plt.show()
 
     """
+    return _normplot('boxcox', x, la, lb, plot, N)
+
+
+def yeojohnson(x, lmbda=None):
+    r"""
+    Return a dataset transformed by a Yeo-Johnson power transformation.
+
+    Parameters
+    ----------
+    x : ndarray
+        Input array.  Should be 1-dimensional.
+    lmbda : {None, scalar}, optional
+        If `lmbda` is not None, do the transformation for that value.
+
+        If `lmbda` is None, find the lambda that maximizes the log-likelihood
+        function and return it as the second output argument.
+
+    Returns
+    -------
+    boxcox : ndarray
+        Box-Cox power transformed array.
+    maxlog : float, optional
```

Comment:
We are explicitly avoiding varying number of outputs (cf. matlab). If it is cheap enough which seems to be the case here return it every time.

Reply:
I think it's best to keep the API like this, as this is what boxcox is doing as well.

Reply:
If it is a new function we don't want to increase the number of outliers but decrease them.

Reply:
I can really see users experimenting with boxcox and yeojohnson in a generic manner, having a different signature complicates their code.

But OK.

I think that if lmbda is not None, the second returned value should be lmbda, or maybe None. Computing the best lambda every time is not that cheap:

```
In [2]: x =  stats.loggamma.rvs(5, size=1000000)                                                                                                                                               

In [4]: %timeit stats.yeojohnson(x, lmbda=None)                                                                                                                                                
2.16 s ± 1.93 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit stats.yeojohnson(x, lmbda=1.5)                                                                                                                                                 
71.6 ms ± 142 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Reply:
Hmm, let's call in some fresh eyes maybe I'm a bit too opinionated about this due to my matlab past.

@chrisb83 @josef-pkt @ev-br and others, do you mind weighing in for the variable number of output issue?

Reply:
> I can really see users experimenting with boxcox and yeojohnson in a generic manner, having a different signature complicates their code.

I think that is a fair point, so I think keeping the API like it is for consistency can be justified.

Reply:
The best would be probably to actually implement bunches and use them in both functions 
(original issue: https://github.com/scipy/scipy/issues/3665; pretty much PR ready class: https://github.com/scipy/scipy/pull/6335). But this could/should be done in a follow-up PR anyway.

Assuming we're not getting Bunches, I do suggest to think again about the optional returns.
(While I don't want to be a pain in the neck) All in all I'm -0 on optional returns.

While consistency with boxcox is indeed desirable, can it be achieved by always calling  `xt, _ = stats.boxcox(x)` and likewise for `yeojohnson`?

Reply:
>While consistency with boxcox is indeed desirable, can it be achieved by always calling xt, _ = stats.boxcox(x) and likewise for yeojohnson?

If you only call `boxcox(x)` or `yeojohnson(x)` yes, but the problem is when you're passing a non-None `lmbda`:

```python
def yeojohnson_wrapper(x, lmbda):
    """Just a wrapper that makes yeo_johnson always return a tuple
    (proposed change)"""
    ret = stats.yeojohnson(x, lmbda)
    if not isinstance(ret, tuple):
        return ret, lmbda
    return ret

np.random.seed(54321)
x = stats.norm.rvs(size=100, loc=100)

for f in stats.yeojohnson, stats.boxcox:
    x_trans, lmbda_max = f(x, lmbda=None)  # works all the time
    x_trans = f(x, lmbda=3)  # works all the time

for f in yeojohnson_wrapper, stats.boxcox:
    x_trans, lmbda_max = f(x, lmbda=None)  # works for both
    x_trans, lmbda_max = f(x, lmbda=3)  # fails for boxcox that only return x_trans
```

To make it work, one would need to do something like this:
```python
for f in yeojohnson_wrapper, stats.boxcox:
    ret = f(x, lmbda=3)
    if f is yeojohnson_wrapper:
        x_trans = ret[0]
    else:
        x_trans = ret
```

I believe returning a Bunch is a great idea and would solve all problems.  Indeed it should probably be done in another PR.

Reply:
Yup, this is convincing :-). Let's keep the API consistent in thus PR. 

A separate PR fir bunches in both functions would be good, but that's OT.

Reply:
makes sense to me.